extras/src/main/scala/spire/math/extras/interval/IntervalSeq.scala

```diff
@@ -0,0 +1,687 @@
+
+import spire.algebra.{Bool, Eq, Order}
+import spire.math.interval._
+import spire.math.{Interval, Rational}
+
+import scala.annotation.tailrec
+import scala.language.implicitConversions
+import scala.reflect.ClassTag
+
+final class IntervalSeq[T] private (
+    val belowAll: Boolean,
+    private val values: Array[T],
+    private val kinds: Array[Byte],
+    private implicit val order: Order[T]) extends IntervalSet[T, IntervalSeq[T]] { lhs =>
+
+  import IntervalSeq._
+
+  private def binarySearch(key: T): Int = {
+    var low = 0
+    var high = values.length - 1
+    while (low <= high) {
+      val mid = (low + high) >>> 1
+      val midVal = values(mid)
+      val c = order.compare(midVal, key)
+      if (c < 0) {
+        low = mid + 1
+      }
+      else if (c > 0) {
+        high = mid - 1
+      }
+      else {
+        // scalastyle:off return
+        return mid
+        // scalastyle:on return
+      }
+    }
+    -(low + 1)
+  }
+
+  def below(index: Int): Boolean = {
+    if (index == 0)
+      belowAll
+    else
+      valueAbove(kinds(index - 1))
+  }
+
+  def at(value: T): Boolean = {
+    val index = binarySearch(value)
+    if (index >= 0)
+      valueAt(kinds(index))
+    else
+      below(-index - 1)
+  }
+
+  def above(value: T): Boolean = {
+    val index = binarySearch(value)
+    if (index >= 0)
+      valueAbove(kinds(index))
+    else
+      below(-index - 1)
+  }
+
+  def below(value: T): Boolean = {
+    val index = binarySearch(value)
+    if (index > 0)
+      valueAbove(kinds(index - 1))
+    else if (index == 0)
+      belowAll
+    else
+      below(-index - 1)
+  }
+
+  def apply(value:T) = at(value)
+
+  def aboveAll = if (values.isEmpty) belowAll else valueAbove(kinds.last)
+
+  def unary_~ :IntervalSeq[T] = copy(belowAll = !belowAll, kinds = negateKinds(kinds))
+
+  def |(rhs:IntervalSeq[T]) = new Or[T](lhs, rhs).result
+
+  def &(rhs:IntervalSeq[T]) = new And[T](lhs, rhs).result
+
+  def ^(rhs:IntervalSeq[T]) = new Xor[T](lhs, rhs).result
+
+  def intersects(rhs: IntervalSeq[T]): Boolean = !new Disjoint[T](lhs, rhs).result
+
+  def isSupersetOf(rhs: IntervalSeq[T]): Boolean = new IsSupersetOf[T](lhs, rhs).result
+
+  def isProperSupersetOf(rhs: IntervalSeq[T]): Boolean = isSupersetOf(rhs) && (lhs != rhs)
+
+  private def copy(belowAll:Boolean = belowAll, values:Array[T] = values, kinds:Array[Byte] = kinds) =
+    new IntervalSeq[T](belowAll, values, kinds, order)
+
+  override def toString = {
+    if(isEmpty)
+      Interval.empty[T].toString()
+    else
+      intervals.mkString(";")
+  }
+
+  override def hashCode = {
+    belowAll.## * 41 + Arrays.hashCode(kinds) * 23 + Arrays.hashCode(values.asInstanceOf[Array[AnyRef]])
+  }
+
+  override def equals(rhs:Any) = rhs match {
+    case rhs:IntervalSeq[_] =>
+      lhs.belowAll == rhs.belowAll &&
+        Arrays.equals(lhs.kinds, rhs.kinds) &&
+        Arrays.equals(values.asInstanceOf[Array[AnyRef]], rhs.values.asInstanceOf[Array[AnyRef]])
```

Comment:
I'm kind of surprised that this cast works. An array of primitives should throw a `ClassCastException` when you coerce it to an `Array[Object]`. Does this get elided by the scala compiler or something?